File: app-android/src/main/java/com/ustadmobile/port/android/view/LocationListRecyclerAdapter.java
package com.ustadmobile.port.android.view;

import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.app.Activity;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.LocationListPresenter;

import com.ustadmobile.lib.db.entities.LocationWithSubLocationCount;

public class LocationListRecyclerAdapter extends
        PagedListAdapter<LocationWithSubLocationCount,
                LocationListRecyclerAdapter.LocationListViewHolder> {

    Context theContext;
    Activity theActivity;
    LocationListPresenter mPresenter;

    @NonNull
    @Override
    public LocationListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {


        View list = LayoutInflater.from(theContext).inflate(
                R.layout.item_title_with_desc_and_dots, parent, false);
        return new LocationListViewHolder(list);

    }

    @Override
    public void onBindViewHolder(@NonNull LocationListViewHolder holder, int position) {

        LocationWithSubLocationCount entity = getItem(position);

        TextView title = holder.itemView.findViewById(R.id.item_title_with_desc_and_dots_title);
        TextView desc = holder.itemView.findViewById(R.id.item_title_with_desc_and_dots_desc);
        AppCompatImageView menu =
                holder.itemView.findViewById(R.id.item_title_with_desc_and_dots_dots);
        //Options to Edit/Delete every schedule in the list
        menu.setOnClickListener((View v) -> {
            //creating a popup menu
            PopupMenu popup = new PopupMenu(theActivity.getApplicationContext(), v);

            popup.setOnMenuItemClickListener(item -> {
                int i = item.getItemId();
                if (i == R.id.edit) {

                    return true;
                } else if (i == R.id.delete) {

                    return true;
                } else {
                    return false;
                }
            });
            //inflating menu from xml resource
            popup.inflate(R.menu.menu_item_schedule);

            //displaying the popup
            popup.show();
        });


    }

    protected class LocationListViewHolder extends RecyclerView.ViewHolder {
        protected LocationListViewHolder(View itemView) {
            super(itemView);
        }
    }

    protected LocationListRecyclerAdapter(
            @NonNull DiffUtil.ItemCallback<LocationWithSubLocationCount> diffCallback,
            LocationListPresenter thePresenter,
            Context context) {
        super(diffCallback);
        mPresenter = thePresenter;
        theContext = context;
    }


}

File: app-android/src/main/java/com/ustadmobile/port/android/view/LocationListRecyclerAdapter.kt

package com.ustadmobile.port.android.view


import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.toughra.ustadmobile.databinding.ItemLocationListBinding
import com.ustadmobile.core.controller.LocationListItemListener
import com.ustadmobile.lib.db.entities.Location

import com.ustadmobile.port.android.view.util.SelectablePagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import com.ustadmobile.port.android.view.ext.setSelectedIfInList


class LocationListRecyclerAdapter(var itemListener: LocationListItemListener?): SelectablePagedListAdapter<Location, LocationListRecyclerAdapter.LocationListViewHolder>(DIFF_CALLBACK) {

    class LocationListViewHolder(val itemBinding: ItemLocationListBinding): RecyclerView.ViewHolder(itemBinding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LocationListViewHolder {
        val itemBinding = ItemLocationListBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        itemBinding.itemListener = itemListener
        itemBinding.selectablePagedListAdapter = this
        return LocationListViewHolder(itemBinding)
    }

    override fun onBindViewHolder(holder: LocationListViewHolder, position: Int) {
        val item = getItem(position)
        holder.itemBinding.location = item
        holder.itemView.setSelectedIfInList(item, selectedItems, DIFF_CALLBACK)
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        itemListener = null
    }

    companion object {
        val DIFF_CALLBACK: DiffUtil.ItemCallback<Location> = object
            : DiffUtil.ItemCallback<Location>() {
            override fun areItemsTheSame(oldItem: Location,
                                         newItem: Location): Boolean {
                TODO("e.g. insert primary keys here return oldItem.location == newItem.location")
            }

            override fun areContentsTheSame(oldItem: Location,
                                            newItem: Location): Boolean {
                return oldItem == newItem
            }
        }
    }

}